java/source/infodynamics/measures/continuous/kraskov/SInfoCalculatorKraskov.java

```diff
@@ -0,0 +1,146 @@
+ * set of observations, using Kraskov-Stoegbauer-Grassberger (KSG) estimation
+ * (see Kraskov et al., below).</p>
+ *
+ * <p>Usage is as per the paradigm outlined for
+ * {@link MultiVariateInfoMeasureCalculatorCommon}.</p>
+ *
+ * <p>Finally, note that {@link Cloneable} is implemented allowing clone()
+ *  to produce only an automatic shallow copy, which is fine
+ *  for the statistical significance calculation it is intended for
+ *  (none of the array 
+ *  data will be changed there).
+ * </p>
+ *
+ * <p><b>References:</b><br/>
+ * <ul>
+ *  <li>Rosas, F., Mediano, P., Gastpar, M., Jensen, H.,
+ *   "Quantifying high-order effects via multivariate extensions of the
+ *   mutual information".</li>
+ *
+ *  <li>Kraskov, A., Stoegbauer, H., Grassberger, P.,
+ *   <a href="http://dx.doi.org/10.1103/PhysRevE.69.066138">"Estimating mutual information"</a>,
+ *   Physical Review E 69, (2004) 066138.</li>
+ * </ul>
+ *
+ * @author Pedro A.M. Mediano (<a href="pmediano at pm.me">email</a>,
+ * <a href="http://www.doc.ic.ac.uk/~pam213">www</a>)
+ */
+public class SInfoCalculatorKraskov
+  extends MultiVariateInfoMeasureCalculatorKraskov
+  implements Cloneable { // See comments on clonability above
+
+
+  protected double[] partialComputeFromObservations(
+      int startTimePoint, int numTimePoints, boolean returnLocals) throws Exception {
+
+    double startTime = Calendar.getInstance().getTimeInMillis();
+
+    double[] localMi = null;
+    if (returnLocals) {
+      localMi = new double[numTimePoints];
+    }
+    
+    // Constants:
+    double dimensionsMinus1TimesDiGammaN = (double) (dimensions - 1) * digammaN;
+
+    // Count the average number of points within eps_x for each marginal x of each point
+    double totalSumF = 0.0;
+        
+    for (int t = startTimePoint; t < startTimePoint + numTimePoints; t++) {
+      // Compute eps for this time step by
+      //  finding the kth closest neighbour for point t:
+      PriorityQueue<NeighbourNodeData> nnPQ =
+          kdTreeJoint.findKNearestNeighbours(k, t, dynCorrExclTime);
+      // First element in the PQ is the kth NN,
+      //  and epsilon = kthNnData.distance
+      NeighbourNodeData kthNnData = nnPQ.poll();
+
+      // Distance to kth neighbour in joint space
+      double eps = kthNnData.distance;
+
+      double sumF = 0.0;
+
+      sumF += (digammaK - digammaN);
+      
+      for (int d = 0; d < dimensions; d++) {
+        int n_small = rangeSearchersInSmallMarginals[d].countPointsStrictlyWithinR(
+                    t, eps, dynCorrExclTime);
+
+        int n_big = rangeSearchersInBigMarginals[d].countPointsStrictlyWithinR(
+                    t, eps, dynCorrExclTime);
+
+
+        sumF -= (MathsUtils.digamma(n_big + 1) - digammaN)/dimensions;
+        sumF -= (MathsUtils.digamma(n_small + 1) - digammaN)/dimensions;
+
+        if (debug) {
+          // Only tracking this for debugging purposes: 
+          System.out.printf("t=%d, d=%d, n_small=%d, n_big=%d, sumF=%.3f\n",

```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.